### ISSUE TITLE
Add support for Class fields

### ISSUE THREAD

**Issue by fluffynuts** [trullock/NUglify#235]
I'm not sure if this _should_ be supported yet - MDN refers to the feature as experimental, but supported in every browser except IE for a very long time (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes) so feel free to say it's simply not supported yet.

The following test illustrates:

```csharp
[Test]
public void ShouldHandleMemberFields()
{
    // Arrange
    var script = @"
class Moo {
	_someField;
	constructor() {
		this._someField = 123;
	}
}
";
    // Act
    var result = Uglify.Js(script, new CodeSettings()
    {
        LocalRenaming = NUglify.JavaScript.LocalRenaming.KeepAll,
        ScriptVersion = ScriptVersion.EcmaScript6
    });
    // Assert

    Console.Error.WriteLine(result.Errors.JoinWith(","));
    Assert.That(result.HasErrors, Is.False);
}
```

I'm only raising because this became an issue in prod, I think after I updated NUglify after the much-appreciated patch to support the spread operator. I'm not sure why it (the page this was on) was working before and not after though - version change was a minor increment. I do recall seeing before that scripts which were failing to minify would be output as they originally were with errors printed at the top, so perhaps this wasn't being minified at all before.

**Comment by trullock** [comment]
This should work, I'll investigate later

**Comment by trullock** [comment]
Only Stage4 TC39 standards are supported.

It seems reasonable to support this though, as its widely supported by browsers and unlikely to change much, especially from a minification point of view. PRs welcome!

**Comment by trullock** [comment]
We should support static fields whilst we're at it

**Comment by Gruski** [comment]
Yep we need Class fields and static fields and methods. Can't use this library until this is supported.

**Comment by mirrortom** [comment]
I have this situation today , please.

**Comment by trullock** [comment]
I'm afraid I don't have time to work on this right now. But if someone wants to give it a go I can point you in the right directions

**Comment by mirrortom** [comment]
> I'm afraid I don't have time to work on this right now. But if someone wants to give it a go I can point you in the right directions

Thank you for your reply. I abandoned the idea and avoided using class variables. At present, my level is not enough to join the construction of this project.

**Comment by fluffynuts** [comment]
@trullock if it's not too much effort, could you put those pointers in this thread? I might have time to take a crack at it.

**Comment by trullock** [comment]
Disclaimer: This is a starting point and not necessarily a 100% (or even 1%) correct description of what actually needs doing. Its just where I got to in my thinking....

This bit is true, there are several bits to change: The Parser/Scanner, and then the various Analysers which manipulate the parsed AST.

`ParseClassNode` and the associated `ParseClassElement` inside `JSParser` dont understand class fields, as such it just returns some attempted AST. So this is definitely the starting point.

Once the parser understands the fields, it needs to return an appropriate C# representation for these - Im not sure at this stage if any of the existing derived `AstNode`s are appropriate for this, or if we need a new one.

Every implementation of `IVisitor` needs its `void Visit(ClassNode node);` reviewing and possible modifying to then handle the fields, depending on how previous steps were implemented

**Comment by trullock** [comment]
Ive made a `bug_235` branch which starts this process

**Comment by trullock** [comment]
also see the fix for #256 , this needs check needs extending wrt this branch

**Comment by fluffynuts** [comment]
I may get time to look at this at the end of next week, perhaps before if time opens up (:

**Comment by trullock** [comment]
I'll take a look too later and try and add some more notes

**Comment by trullock** [comment]
Ive pushed to changes to https://github.com/trullock/NUglify/tree/bug_235
The Parser changes are almost there, the real question is what to do with the `ClassField` class I've added. Should this exist? Is there an existing class which should really be used instead - if so, which?
Every `IVisitor.Visit(ClassField)` method needs reviewing to see what it needs to do

**Comment by fluffynuts** [comment]
@trullock I'm having a look at this today, but not sure exactly what I'm missing - I can't build - NUglify.Tests appears to be missing `packages.config`? Not sure if I'm missing something else too? I'm going to see if I can work around missing stuff, but would appreciate it if you checked your side too (:

**Comment by trullock** [comment]
you need to do this: https://stackoverflow.com/questions/3796927/how-to-git-clone-including-submodules

**Comment by trullock** [comment]
you dont need to "install" anything, just restore the Nuget packages

**Comment by trullock** [comment]
> you need to do this: https://stackoverflow.com/questions/3796927/how-to-git-clone-including-submodules

This is something to remove the need for IMO, legacy from previous maintainer

**Comment by fluffynuts** [comment]
> you need to do this: [stackoverflow.com/questions/3796927/how-to-git-clone-including-submodules](https://stackoverflow.com/questions/3796927/how-to-git-clone-including-submodules)

O, silly me, I didn't even think of git submodules, even though I use them quite a bit in my own stuff 😁

**Comment by fluffynuts** [comment]
> you dont need to "install" anything, just restore the Nuget packages

Ah, I've done a manual `nuget restore NUglify.sln` and that works. It's been a very long time since I've seen a `project.json` and I've been sitting here for a while wondering how nunit refs worked considering the .csproj doesn't have `PackageReference` nodes and there's no `packages.config` 😆

**Comment by fluffynuts** [comment]
update: NUglify is (as I'm sure you're aware) non-trivial. I'm still wrapping my head around how simple things work, using the `ClassDefinition` test to step-through. I'll hopefully pick this up again next week, but for now, I've only been stepping and reading - I have accomplished exactly nothing.

**Comment by trullock** [comment]
Learning isn't nothing :)

**Comment by trullock** [comment]
Ensure we support:

Class fields, static and not static
Computed names
I suspect getters and setters have a bug with computed names, look at fixing that at the same time, or spin off another ticket

**Comment by trullock** [comment]
@fluffynuts @mirrortom @poulBak @rochapablo Check out v1.15.0-beta on nuget prerelease, I belive it fixes this although I fully expect there are some bugs. Please try and find them and report back providing minimum repro steps to whatevers wrong.

**Comment by PoulBak** [comment]
@trullock: I have just tested the pre release version with a file, that contains classes and static properties. 
It looks fine to me!

Btw: I actually have a 'prettyrizer' (doing the opposite of NUglify) on my website: [https://bak-o-soft.dk/Prettyrizer/Prettyrizer.aspx](https://bak-o-soft.dk/Prettyrizer/Prettyrizer.aspx). Feel free to use it anyway you like.

Poul Bak

**Comment by mirrortom** [comment]
So, Then, this function is about to be realized. Thanks author!

**Comment by trullock** [comment]
Well, im going to merge this into master.

When you find something broken, open a new ticket

**Comment by trullock** [comment]
released as 1.15.0

**Comment by fluffynuts** [comment]
Thanks for your hard work @trullock
